File: leetcode/medium/generate-parentheses.cpp
// https://leetcode.com/problems/generate-parentheses/
//
// Uses backtracking.
// Complexity: complexity of generating n Catalan numbers: runtime
// O(4^n/sqrt(n)), space O(n+(4^n/sqrt(n))) [from the leetcode analysis]

class Solution {
    void generate(vector<string>& out, string& combination,
                  const int numRemaining, const int numUnclosed = 0) const
    {
        if (numRemaining < 0 || numUnclosed < 0)
            return; // ill-formed parens in combination

        if (numRemaining == 0 && numUnclosed == 0) {
            out.push_back(combination);
            return; // completed a valid combination
        }

        generate(out, combination += '(', numRemaining - 1, numUnclosed + 1);
        combination.back() = ')'; // backtrack the '(', replace with ')'
        generate(out, combination, numRemaining, numUnclosed - 1);
        combination.pop_back(); // backtrack the ')' for our caller
    }

public:
    vector<string> generateParenthesis(const int n) const
    {
        vector<string> results;
        string combination;

        generate(results, combination, n);
        return results;
    }
};

File: leetcode/medium/generate-parentheses.rs
// https://leetcode.com/problems/generate-parentheses
//
// Complexity: see C++ solution.

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let (mut res, mut tmp) = (vec![], String::with_capacity(n as usize * 2));
        Self::generate(&mut res, &mut tmp, n as _, 0);
        res
    }

    fn generate(out: &mut Vec<String>, cur: &mut String, n: u32, open: u32) {
        if (n, open) == (0, 0) {
            out.push(cur.clone());
        } else {
            let mut backtrack = |c, n, open, cond| {
                if cond {
                    cur.push(c);
                    Self::generate(out, cur, n, open);
                    cur.pop();
                }
            };
            backtrack('(', n - 1, open + 1, n > 0);
            backtrack(')', n, open - 1, open > 0);
        }
    }
}
